docs/proposals/config_data.md

```diff
@@ -0,0 +1,297 @@
+  Key            string `json:"key"`
+}
+```
+
+### Volume Source
+
+A new `ConfigDataVolumeSource` type of volume source containing the `ConfigData` object will be
+added to the `VolumeSource` struct in the API:
+
+```go
+package api
+
+type VolumeSource struct {
+  // other fields omitted
+  ConfigData *ConfigDataVolumeSource `json:"configData,omitempty"`
+}
+
+type ConfigDataVolumeSource struct {
+  Items []ConfigDataVolumeFile `json:"items,omitempty"`
+}
+
+type ConfigDataVolumeFile struct {
+  Path       string             `json:"path"`
+  ConfigData ConfigDataSelector `json:"configData"`
+}
+
+
+```
+
+**Note:** The update logic used in the downward API volume plug-in will be extracted and re-used in
+the volume plug-in for `ConfigData`.
+
+## Examples
+
+#### Consuming ConfigData as environment variables
+
+```yaml
+apiVersion: extensions/v1beta1
+name: etcd-env-config
+kind: ConfigData
+data:
+  number_of_members: 1
+  initial_cluster_state: new
+  initial_cluster_token: DUMMY_ETCD_INITIAL_CLUSTER_TOKEN
+  discovery_token: DUMMY_ETCD_DISCOVERY_TOKEN
+  discovery_url: http://etcd-discovery:2379
+  etcdctl_peers: http://etcd:2379
+```
+
+This pod consumes the `ConfigData` as environment variables:
+
+```yaml
+apiVersion: v1
+kind: Pod
+spec:
+  containers:
+  - name: etcd
+    image: openshift/etcd-20-centos7
+    ports:
+    - containerPort: 2379
+      protocol: TCP
+    - containerPort: 2380
+      protocol: TCP
+    env:
+    - name: ETCD_NUM_MEMBERS
+      valueFrom:
+        configData:
+          configDataName: etcd-env-config
+          key: num_members
+    - name: ETCD_INITIAL_CLUSTER_STATE
+      valueFrom:
+        configData:
+          configDataName: etcd-env-config
+          key: initial_cluster_state
+    - name: ETCD_DISCOVERY_TOKEN
+      valueFrom:
+        configData:
+          configDataName: etcd-env-config
+          key: dicovery_token
+    - name: ETCD_DISCOVERY_URL
+      valueFrom:
+        configData:
+          configDataName: etcd-env-config
+          key: discovery_url
+    - name: ETCDCTL_PEERS
+      valueFrom:
+        configData:
+          configDataName: etcd-env-config
+          key: etcdctl_peers
+```
+
+### Consuming `ConfigData` as volumes
+
+`redis-volume-config` is intended to be used as a volume containing a config file:
+
+```yaml
+apiVersion: extensions/v1beta1
+name: redis-volume-config
+kind: ConfigData
+data:
+  redis.conf: "pidfile /var/run/redis.pid\nport6379\ntcp-backlog 511\n databases 1\ntimeout 0\n"
+```
+
+The following pod consumes the `redis-volume-config` in a volume:
+
+```yaml
+apiVersion: v1
+kind: Pod
+name: example-volume-pod
+spec:
+  containers:
+    - name: redis
+      image: kubernetes/redis
+      command: "redis-server /mnt/config-data/etc/redis.conf"
+      ports:
+        - containerPort: 6379
+      volumeMounts:
+        - name: config-data-volume
+          mountPath: /mnt/config-data
+  volumes:
+  - name: config-data-volume
+    configDataVolumes:
```

Comment:
This doesn't match the types above.

I do see what the path is for, though it does seem like multiple volumes could work.